content/algorithms/flow/dinitz_alg.md

```diff
@@ -0,0 +1,687 @@
+```
+
+```
+gif("images/example",limage,fps=0.5)
+```
+````
+
+```{code-cell} ipython3
+%matplotlib inline
+import networkx as nx
+import matplotlib.pyplot as plt
+import pickle
+```
+
+```{code-cell} ipython3
+gname = "egnetwork"
+# loading the graph
+G = nx.read_graphml(f"data/{gname}.graphml")
+with open(f"data/pos_{gname}", "rb") as fp:
+    pos = pickle.load(fp)
+```
+
+```{code-cell} ipython3
+fig, axes = plt.subplots(4, 2, figsize=(20, 30))
+
+# assign colors and labels to nodes based on their type
+color_map = {"t": "skyblue", "s": "skyblue"}
+node_colors = [color_map[u] if u in color_map.keys() else "0.8" for u in G.nodes]
+node_labels = {u: u for u in G.nodes}
+cutoff_list = [5, 10, 15, 20, 25, 30, 35, 40]
+
+for i in range(8):
+
+    # calculating the maximum flow with the cutoff value
+    R = nx.flow.dinitz(G, s="s", t="t", capacity="capacity", cutoff=cutoff_list[i])
+
+    # coloring and labeling edges depending on if they have non-zero flow value or not
+    edge_colors = ["0.8" if R[u][v]["flow"] == 0 else "0" for u, v in G.edges]
+    edge_labels = {
+        (u, v): f"{R[u][v]['flow']}/{G[u][v]['capacity']}"
+        for u, v in G.edges
+        if R[u][v]["flow"] != 0
+    }
+
+    # drawing the network
+    nx.draw_networkx_nodes(
+        G, pos=pos, ax=axes[i // 2][i % 2], node_size=500, node_color=node_colors
+    )
+    nx.draw_networkx_labels(
+        G, pos=pos, ax=axes[i // 2][i % 2], labels=node_labels, font_size=14
+    )
+    nx.draw_networkx_edges(G, pos=pos, ax=axes[i // 2][i % 2], edge_color=edge_colors)
+    nx.draw_networkx_edge_labels(
+        G, pos=pos, ax=axes[i // 2][i % 2], edge_labels=edge_labels, font_size=14
+    )
+    axes[i // 2][i % 2].set_title(
+        f"Max Flow = {R.graph['flow_value']}\nCutoff value of = {cutoff_list[i]}",
+        size=15,
+    )
+
+fig.tight_layout()
+```
+
+Note: Iteration are stopped if the maximum flow found so far exceeds the cutoff value
+## Reductions and Applications
+There are many other problems which can be reduced to Maximum flow problem for eg.
+* [Maximum Bipartite Matching](https://en.wikipedia.org/wiki/Matching_(graph_theory))
+* [Assignment Problem](https://en.wikipedia.org/wiki/Assignment_problem)
+* [Transportation Problem](https://en.wikipedia.org/wiki/Transportation_theory_(mathematics))
+
+and many others
+
+Note that even though dinitz works in $O(n^2m)$ strongly polynomial time, i.e. to say it 
+doesn't depend on the value of flow. It is noteworthy that its performance of biparted 
+graphs is especially fast being $O(\sqrt n m)$ time, where $n = |V|$ & $m = |E|$.
+
+Lets consider the example of shipping packages from warehouse to customers through some 
+intermediate shipping points, and we can only ship limited number of packages through 
+an intermediate shipping point in a day.
+
+So how to assign intermediate shipping point to customer so that maximum number of 
+packages are shipped in a day?
```

Comment:
```suggestion
Given these constraints, the shipping company would like to assign intermediate shipping points to customers so that the maximum number of packages are shipped each day.
```